Assets/Tests/Runtime/NetworkedMessageChannelTests.cs

```diff
@@ -0,0 +1,204 @@
+            m_Subscriptions.Add(genericMessageChannelClient.Subscribe(OnGenericMessageReceived));
+
+            void OnGenericMessageReceived(GenericMessage message)
+            {
+                nbMessagesReceived++;
+                Assert.IsTrue(message.value);
+            }
+
+            genericMessageChannelServer.Publish(new GenericMessage() { value = true });
+
+            // wait for the custom named message to be sent on the server and received on the client
+            yield return null;
+            yield return null;
+
+            Assert.AreEqual(1, nbMessagesReceived);
+        }
+
+        [UnityTest]
+        public IEnumerator NetworkedMessagesAreStillReceivedAfterNetworkManagerShutsDownAndRestarts()
+        {
+            var emptyMessageChannelClient1 = new NetworkedMessageChannel<EmptyMessage>();
+            m_FirstClientScope.InjectIn(emptyMessageChannelClient1);
+            var emptyMessageChannelClient2 = new NetworkedMessageChannel<EmptyMessage>();
+            m_SecondClientScope.InjectIn(emptyMessageChannelClient2);
+            var emptyMessageChannelServer = new NetworkedMessageChannel<EmptyMessage>();
+            m_ServerScope.InjectIn(emptyMessageChannelServer);
+
+            var nbMessagesReceived = 0;
+
+            m_Subscriptions = new DisposableGroup();
+            m_Subscriptions.Add(emptyMessageChannelClient1.Subscribe(OnEmptyMessageReceived));
+            m_Subscriptions.Add(emptyMessageChannelClient2.Subscribe(OnEmptyMessageReceived));
+
+            void OnEmptyMessageReceived(EmptyMessage message)
+            {
+                nbMessagesReceived++;
+            }
+
+            emptyMessageChannelServer.Publish(new EmptyMessage());
+
+            // wait for the custom named message to be sent on the server and received on the clients
+            yield return null;
+            yield return null;
+
+            Assert.AreEqual(2, nbMessagesReceived);
+
+            // Shutdown the server and clients
+            m_ServerNetworkManager.Shutdown();
+            FirstClient.Shutdown();
+            SecondClient.Shutdown();
+
+            yield return new WaitWhile(() => m_ServerNetworkManager.ShutdownInProgress);
+            yield return new WaitWhile(() => FirstClient.ShutdownInProgress);
+            yield return new WaitWhile(() => SecondClient.ShutdownInProgress);
+
+            // Restart the server and clients
+            m_ServerNetworkManager.StartServer();
+            FirstClient.StartClient();
+            SecondClient.StartClient();
+
+            yield return WaitForClientsConnectedOrTimeOut();
+
+            // Test sending a message a second time
+            emptyMessageChannelServer.Publish(new EmptyMessage());
+
+            // wait for the custom named message to be sent on the server and received on the clients
+            yield return null;
+            yield return null;
+
+            Assert.AreEqual(4, nbMessagesReceived);
```

Comment:
to make this clearer, I'd just set nbMessageReceived to 0 before the send, then check for 2 again.

Reply:
When I did, Rider gave me a warning in OnEmptyMessageReceived because the captured variable is updated in the outer scope, but since this won't actually cause issues, I could ignore the warning

Reply:
yeah, silly rider

Reply:
wait, why is it a class property now?

Reply:
I moved it as a field so that I could extract the code creating the message channels and the subscriptions into a reusable method to reduce code duplication

Reply:
as soon as you add persistent state to your tests, you risk having spreading corruption between test runs. Each test should have its own isolated world as much as possible. Either this or you have setup and teardown methods that'll make sure they run even when a test fails (an assert that fails throws an exception, interrupting the test and not running any teardown you have in the actual test, you need to have teardown methods to make sure they still run at the end)

Reply:
Yes, that is why that field is set to 0 on Setup